File: min_stack.py
"""class MinStack:
    ordinary=[]
    minimium=[]
    def __init__(self):
        self.ordinary = []
        self.minimium = []
    def push(self,x):
        #ordinary.append(x)
        self.ordinary.append(x)
        if len(self.minimium) == 0:
            self.minimium.append(x)       
        else:
            self.minimium.append(min(x,self.minimium[-1]))
        
        return self.minimium[-1]
    def pop(self):
        if self.minimium is not []:
            self.minimium.pop()
            self.ordinary.pop()

    def top(self):
        if self.ordinary is []:
            return -1
        return self.ordinary[-1]
    def getMin(self):
        if self.minimium is []:
            return -1
        return self.minimium[-1]
    """
class MinStack:
    # @param x, an integer
    # @return an integer
    stack, minStack = [], []
    global stack, minStack
    def push(self, x):
        stack.append(x)
        if not minStack or minStack[-1] >= x:
            minStack.append(x)
        return minStack[-1]
    # @return nothing
    def pop(self):
        if not stack:
            return
        if stack.pop() == minStack[-1]:
            minStack.pop()
    # @return an integer
    def top(self):
        return stack[-1] if  stack else -1

    # @return an integer
    def getMin(self):
        return minStack[-1] if minStack else -1
if __name__=='__main__':
    sol=MinStack()
    lst=[2,3,4,7,4,8,3,6,4,1,6]
    for element in lst:
        sol.push(element)
    print sol.top()
    sol.pop()
    sol.pop()
    print sol.getMin()

File: Algorithm/CPP/Min_Stack.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <unordered_map>
# include <cctype>
# include <algorithm>
# include <time.h>
using namespace std;



class MinStack {
public:
	void push(int x) {
		Stack.push(x);
		if (Min.empty() || Min.top() >= x)
			Min.push(x);

	}

	void pop() {
		int top = Stack.top();
		Stack.pop(); 
		if (top == Min.top())
			Min.pop();
	}

	int top() {
		return Stack.top();
	}

	int getMin() {
		return Min.top();
	}
private:
	stack<int> Stack;
	stack<int> Min;
};
int  main()
{
	MinStack ms;
	ms.push(0);
	ms.push(1);
	ms.push(0);
	//cout << ms.top() << endl;
	cout << ms.getMin() << endl;
	ms.pop();
	//cout << ms.top() << endl;
	cout << ms.getMin() << endl;
	cout << endl;
	system("pause");
	return 0;
}
